compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt

```diff
@@ -338,3 +263,89 @@ class BackendCodeGenerator(
             CodeBlock.of("%N.%M(%L, %N)", "runtime", sendMember, value, receiver.name)
     }
 }
+
+private fun addHostDeclarations(fileBuilder: FileSpec.Builder, program: ProgramNode) {
+    // add a global host object for each host
+    for (host: Host in program.hosts) {
+        fileBuilder.addProperty(
+            PropertySpec.builder(host.name, Host::class)
+                .initializer(
+                    CodeBlock.of(
+                        "%T(%S)",
+                        Host::class,
+                        host.name
+                    )
+                )
+                .build()
+        )
+    }
+}
+
+fun compileKotlinFile(
```

Comment:
why is `compileKotlinFile` and compileKotlinFileSpec` top-level functions and not a part of `BackendCodeGenerator`?

Reply:
`BackendCodeGenerator`s are created per host. These functions are per file. Also, having a function and not a class is the right thing here. Creating a class that you run once is an anti-pattern.